### ISSUE TITLE
Building Julia 0.4 + LLVM svn fails

### ISSUE THREAD

**Issue by SimonDanisch** [JuliaLang/julia#12037]
Hi, 
tried to compile Julia's current master b414076bc4e9b77f983524540f0d8ad9498f1aa1 with LLVM svn in cygwin.
Error:
https://gist.github.com/SimonDanisch/a7713ad1f9ac95237f63#file-builderror
Make.user:
```Make
XC_HOST = x86_64-w64-mingw32
override LLDB_VER=master
override LLVM_VER=svn
override LLVM_ASSERTIONS=1
override BUILD_LLVM_CLANG=1
override BUILD_LLDB=1
override USE_LLVM_SHLIB=1
override LLDB_DISABLE_PYTHON=1
```

Best,
Simon

**Comment by tkelman** [comment]
Huh, looks like LLVM may have added dllexports when building an LLVM dll. Whoever did so probably only tested with MSVC? Can you recall the last time you did a successful Windows build of LLVM-svn, or at least a build that didn't hit these errors? We probably need to complain on llvmdev or phabricator that they broke the mingw-w64 build. If they have mingw-w64 buildbots they may find that out sooner or later.

**Comment by SimonDanisch** [comment]
It has never build successfully :D 
The last time I tried must be ages ago... Can't really tell, sorry...

**Comment by joa-quim** [comment]
I did a build 9 days ago

https://github.com/JuliaLang/julia/issues/11726

**Comment by ihnorton** [comment]
If you are building as a shared library on Windows, you need this:
http://reviews.llvm.org/D5036

**Comment by tkelman** [comment]
Ah, right. A note about that should really be put into the Cxx.jl readme. And we should try bumping it again upstream. Closing since no real action items to do here - we could carry that as a local patch but I suspect other things will be broken with a dll of llvm?

**Comment by ihnorton** [comment]
It bootstraps fine, and I don't think there were any unexpected test
failures.

There are currently some incorrect signatures in debuginfo related to
sym_iter. Keno just submitted a PR about it, but there are more inside the
Windows ifdefs. I'll fix those soon if someone doesn't get to it first.

**Comment by tkelman** [comment]
In that case let's carry that patch then, since who knows if or when it'll ever get merged upstream. Should carefully test which versions of LLVM the same patch file can be made to apply cleanly to.

**Comment by ihnorton** [comment]
Also note that LLDB doesn't currently build under mingw64 due to unsupported C++11 threading/sync constructs.

@Keno: any chance you can get the that llvm patch in?

**Comment by tkelman** [comment]
We could try the posix-threads version of mingw64, but I think that's generally slower and doesn't match what opensuse is using for winrpm packages. Might also need a build system patch or two here to make that version work.

**Comment by ihnorton** [comment]
@SimonDanisch, for reference, here is my `Make.user`, with which I'm able to bootstrap again after f42839200373:

XC_HOST = x86_64-w64-mingw32

```
override LLDB_VER=master
override LLVM_VER=svn
override LLVM_ASSERTIONS=1
override BUILD_LLVM_CLANG=1
#override BUILD_LLDB=1
override USE_LLVM_SHLIB=1
override LLDB_DISABLE_PYTHON=1
```

**Comment by tkelman** [comment]
@ihnorton for my reference, are you cross-compiling from cygwin or following the recent msys2 instructions using the opensuse toolchain?

**Comment by ihnorton** [comment]
Cross-compiling from cygwin (at the moment I don't really use this for anything other than testing/fixing the llvm-svn build on windows, so there might be package issues I don't see).

**Comment by tkelman** [comment]
Cool, thanks. If you haven't run cygwin's setup within the last few weeks, then I would expect package issues. Some time the machine's not busy or before you go to sleep, update the packages, distclean and rebuild llvm and any other c++ deps to get to a consistent ABI with the buildbots.

**Comment by SimonDanisch** [comment]
@ihnorton 
So you apply patch f4283920037399a0ecefa4507adbc6eb9a25708e, fc3df9640990430e55c67f28a4917a5bd84f30b4 and http://reviews.llvm.org/D5036 with the make file you posted and that is all that needs to be done in order to build julia's master with llvm svn and cygwin?
Just trying to make sure before I start another rebuild of LLVM, as I did this and it did not work out with only `make cleanall`.

**Comment by ihnorton** [comment]
Well, you shouldn't have to apply any Julia patches -- just update to
master.

For llvm, I usually reset my branch before pulling llvm-svn, and then
re-cherry-pick a commit applying D5036. I don't think I had to do anything
else.

I just tried updating Cxx.jl and it does not build at the moment due to a
supposed syntax error (maybe a gcc thing, not sure yet), but that is
unrelated to building Julia.

**Comment by joa-quim** [comment]
Hmm, with the msys2 build, it doesn't because

    In file included from E:/julia/deps/llvm-svn/lib/MC/MCELF.cpp:16:0:
    E:/julia/deps/llvm-svn/include/llvm/MC/MCSymbol.h: In static member function 'static void llvm::MCELF::SetBinding(const llvm::MCSymbol&, unsigned int)':
    E:/julia/deps/llvm-svn/include/llvm/MC/MCSymbol.h:403:12: error: 'uint32_t llvm::MCSymbol::getFlags() const' is protected
       uint32_t getFlags() const { return Flags; }
                ^
    E:/julia/deps/llvm-svn/lib/MC/MCELF.cpp:26:38: error: within this context
       uint32_t OtherFlags = Sym.getFlags() & ~(0xf << ELF_STB_Shift);
    ...

but, like my previous attempts, I may have to start clean with LLVM SVN

**Comment by SimonDanisch** [comment]
Ugh... After half a day of compilation:

```
dict.jl
iterator.jl
inference.jl
Declaration may not be in a Comdat!
void ()* @__stack_chk_fail
Declaration may not be in a Comdat!
void (i8*)* @jl_error
Declaration may not be in a Comdat!
void (%jl_value_t*)* @jl_throw
Declaration may not be in a Comdat!
void (%jl_value_t*)* @jl_undefined_var_error
Declaration may not be in a Comdat!
void (%jl_value_t*, i64*, i64)* @jl_bounds_error_ints
Declaration may not be in a Comdat!
void (%jl_value_t*, i64)* @jl_bounds_error_int
Declaration may not be in a Comdat!
void (%jl_value_t**, i64, i64)* @jl_bounds_error_tuple_int
Declaration may not be in a Comdat!
void (i8*, %jl_value_t*, i64)* @jl_bounds_error_unboxed_int
Declaration may not be in a Comdat!
void (%jl_value_t*, i32)* @jl_throw_with_superfluous_argument
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_new_structv
Declaration may not be in a Comdat!
i32 (i8*)* @jl_setjmp
Declaration may not be in a Comdat!
i32 (i8*, i8*, i64)* @memcmp
Declaration may not be in a Comdat!
void (i8*, i8*, %jl_value_t*, %jl_value_t*, i32)* @jl_type_error_rt_line
Declaration may not be in a Comdat!
void (%jl_value_t*, %jl_value_t*)* @jl_checked_assignment
Declaration may not be in a Comdat!
void (%jl_value_t*)* @jl_declare_constant
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t*)* @jl_get_binding_or_error
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_is
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_typeof
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_sizeof
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_subtype
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_isa
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_typeassert
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_apply
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_kwcall
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_throw
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_tuple
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_svec
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_methodexists
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_applicable
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_invoke
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_top_eval
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_isdefined
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_get_field
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_set_field
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_field_type
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_nfields
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_new_expr
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_arraylen
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_arrayref
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_arrayset
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_arraysize
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_f_instantiate_type
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, i32)* @jl_apply_generic
Declaration may not be in a Comdat!
void (%jl_value_t*)* @jl_gc_queue_root
Declaration may not be in a Comdat!
void (%jl_value_t*, %jl_value_t*)* @jl_gc_wb_slow
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*)* @jl_new_box
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*)* @jl_toplevel_eval
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*)* @jl_copy_ast
Declaration may not be in a Comdat!
%jl_value_t* (i8*, %jl_value_t*, %jl_value_t*)* @jl_new_closure
Declaration may not be in a Comdat!
%jl_value_t* (i64, ...)* @jl_svec
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, %jl_value_t*, %jl_value_t*, %jl_value_t*, %jl_value_t*, %jl_value_t*, %jl_value_t*, i32)* @jl_method_def
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, %jl_value_t**, %jl_value_t*, %jl_value_t*)* @jl_generic_function_def
Declaration may not be in a Comdat!
void (i8*)* @jl_enter_handler
Declaration may not be in a Comdat!
void ()* @_resetstkoflw
Declaration may not be in a Comdat!
void ()* @___chkstk_ms
Declaration may not be in a Comdat!
void (i32)* @jl_pop_handler
Declaration may not be in a Comdat!
i32 (%jl_value_t*, %jl_value_t*)* @jl_egal
Declaration may not be in a Comdat!
i32 (%jl_value_t*, %jl_value_t*, i32)* @jl_subtype
Declaration may not be in a Comdat!
%jl_value_t* (i64)* @jl_gc_allocobj
Declaration may not be in a Comdat!
%jl_value_t* ()* @jl_gc_alloc_1w
Declaration may not be in a Comdat!
%jl_value_t* ()* @jl_gc_alloc_2w
Declaration may not be in a Comdat!
%jl_value_t* ()* @jl_gc_alloc_3w
Declaration may not be in a Comdat!
%jl_value_t* (i64)* @jl_alloc_svec
Declaration may not be in a Comdat!
i8* (i8*, i8*, i8**)* @jl_load_and_lookup
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i8*)* @jl_new_bits
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i64)* @jl_get_nth_field_checked
Declaration may not be in a Comdat!
i64 ()* @jl_gc_diff_total_bytes
Declaration may not be in a Comdat!
void (i8*, i32)* @show_execution_point
Declaration may not be in a Comdat!
%jl_value_t* (i8)* @jl_box_int8
Declaration may not be in a Comdat!
%jl_value_t* (i8)* @jl_box_uint8
Declaration may not be in a Comdat!
%jl_value_t* (i16)* @jl_box_int16
Declaration may not be in a Comdat!
%jl_value_t* (i16)* @jl_box_uint16
Declaration may not be in a Comdat!
%jl_value_t* (i32)* @jl_box_int32
Declaration may not be in a Comdat!
%jl_value_t* (i32)* @jl_box_uint32
Declaration may not be in a Comdat!
%jl_value_t* (i64)* @jl_box_int64
Declaration may not be in a Comdat!
%jl_value_t* (i64)* @jl_box_uint64
Declaration may not be in a Comdat!
%jl_value_t* (float)* @jl_box_float32
Declaration may not be in a Comdat!
%jl_value_t* (double)* @jl_box_float64
Declaration may not be in a Comdat!
%jl_value_t* (i32)* @jl_box_char
Declaration may not be in a Comdat!
%jl_value_t* (i64)* @jl_box_gensym
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i8)* @jl_box8
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i16)* @jl_box16
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i32)* @jl_box32
Declaration may not be in a Comdat!
%jl_value_t* (%jl_value_t*, i64)* @jl_box64
Declaration may not be in a Comdat!
%jl_value_t*** @jl_pgcstack
Declaration may not be in a Comdat!
%jl_value_t** @__stack_chk_guard
Declaration may not be in a Comdat!
%jl_value_t** @jl_true
Declaration may not be in a Comdat!
%jl_value_t** @jl_false
Declaration may not be in a Comdat!
%jl_value_t** @jl_emptysvec
Declaration may not be in a Comdat!
%jl_value_t** @jl_emptytuple
Declaration may not be in a Comdat!
%jl_value_t** @jl_exception_in_transit
Declaration may not be in a Comdat!
%jl_value_t** @jl_diverror_exception
Declaration may not be in a Comdat!
%jl_value_t** @jl_undefref_exception
Declaration may not be in a Comdat!
%jl_value_t** @jl_domain_exception
Declaration may not be in a Comdat!
%jl_value_t** @jl_overflow_exception
Declaration may not be in a Comdat!
%jl_value_t** @jl_inexact_exception
Declaration may not be in a Comdat!
i8** @jl_RTLD_DEFAULT_handle
Declaration may not be in a Comdat!
i8** @jl_exe_handle
Declaration may not be in a Comdat!
i8** @jl_dl_handle
LLVM ERROR: Broken module found, compilation aborted!
Makefile:177: recipe for target '/home/S/julia/usr/lib/julia/inference0.o' failed
make[1]: *** [/home/S/julia/usr/lib/julia/inference0.o] Error 1
Makefile:64: recipe for target 'julia-sysimg' failed
make: *** [julia-sysimg] Error 2

```

**Comment by tkelman** [comment]
huh, well, that's new - cc @vtjnash any idea?

**Comment by ihnorton** [comment]
Yes, I just saw that too. The last time I pulled llvm was right before this went in:

https://github.com/llvm-mirror/llvm/commit/cf47c180e1e8f6071448555b3a67220a7e72c38a

**Comment by vtjnash** [comment]
huh, well i guess i need to be a little more careful what goes into the comdat now. before I just stuck everything in the comdat section and assumed that llvm would just figure it out.

but please remember that I really do not have the bandwidth to test llvm-svn fixes. i'll build against it occasionally, but don't have the resources nor interest to keep it continuously functional

**Comment by ihnorton** [comment]
This patch gets us to a REPL:
```
diff --git a/src/cgutils.cpp b/src/cgutils.cpp
index 92d1edd..0af7327 100644
--- a/src/cgutils.cpp
+++ b/src/cgutils.cpp
@@ -6,7 +6,7 @@
 template<class T> // for GlobalObject's
 static T* addComdat(T *G)
 {
-    if (imaging_mode) {
+    if (imaging_mode && (!G->isDeclarationForLinker())) {
         Comdat *jl_Comdat = shadow_module->getOrInsertComdat(G->getName());
         jl_Comdat->setSelectionKind(Comdat::NoDuplicates);
         G->setComdat(jl_Comdat);
```

**Comment by vtjnash** [comment]
lgtm – that's essentially what i was expecting to be needed

**Comment by tkelman** [comment]
Oh that was easy. Sorry @vtjnash I know you can't do too much for llvm svn, I never build against it myself, but the comdat looked relevant.

**Comment by ihnorton** [comment]
https://github.com/JuliaLang/julia/commit/0503f2aeeacecc8bf5cf5b464891cf3507196321

@SimonDanisch fortunately this one you don't have to rebuild llvm :)

**Comment by SimonDanisch** [comment]
Nice! This works with GLVisualize =)
It seems to suffer from a huge compile time regression, though.
Loading GLVisualize went from an unbearable 30 seconds to something like 2-3 minutes!
But the run time seems to be equally fast.

**Comment by tkelman** [comment]
codegen performance regression is known, and one of the things to fix before #9336 can be closed

**Comment by ihnorton** [comment]
(though it does seem to have become worse recently...)
